rpcserver.go

```diff
@@ -6179,18 +6257,92 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
 			return nil, err
 		}
 
+	// There is no need to register a PSBT shim before opening the channel,
+	// even though our RPC message structure allows for it. Inform the user
+	// by returning a proper error instead of just doing nothing.
+	case in.GetShimRegister() != nil &&
+		in.GetShimRegister().GetPsbtShim() != nil:
+
+		return nil, fmt.Errorf("PSBT shim must only be sent when " +
+			"opening a channel")
+
 	// If this is a transition to cancel an existing shim, then we'll pass
 	// this message along to the wallet.
 	case in.GetShimCancel() != nil:
-		pid := in.GetShimCancel().PendingChanId
+		copy(pendingChanID[:], in.GetShimCancel().PendingChanId)
+		rpcsLog.Debugf("Canceling funding shim for pending_id=%x",
+			pendingChanID)
+
+		// We might need to cancel the started negotiation between us
+		// and the peer, depending on the kind of funding flow. The
+		// intent will know what to do on being canceled.
+		intent, err := r.server.cc.wallet.GetFundingIntent(
+			pendingChanID,
+		)
+		if err != nil {
+			return nil, err
+		}
+		intent.Cancel()
+
+		// Also inform the wallet that the intent no longer needs to be
+		// considered.
+		err = r.server.cc.wallet.CancelFundingIntent(pendingChanID)
+		if err != nil {
+			return nil, err
+		}
 
-		var pendingChanID [32]byte
-		copy(pendingChanID[:], pid)
+	// If this is a transition to verify the PSBT for an existing shim,
+	// we'll do so and then store the verified PSBT for later so we can
+	// compare it to the final, signed one.
+	case in.GetPsbtVerify() != nil:
+		copy(pendingChanID[:], in.GetPsbtVerify().PendingChanId)
+		psbtBytes := in.GetPsbtVerify().FundedPsbt
 
-		err := r.server.cc.wallet.CancelFundingIntent(pendingChanID)
+		rpcsLog.Debugf("Verifying PSBT for pending_id=%x",
+			pendingChanID)
+		intent, err := r.server.cc.wallet.GetFundingIntent(
```

Comment:
In self-response to my comment above, perhaps this can be pushed down further into the wallet? Which would then mean we don't really need the `GetFundingIntent` method as the wallet is able to look it up itself by reservation (with the added methods for the reservation being added). Don't feel strongly about either path though in the end.

Reply:
No, you're right. It absolutely makes sense to move this, makes the code much more clean.